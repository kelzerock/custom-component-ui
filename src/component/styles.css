html,
body {
  height: 100%;
  margin: 0;
}

.wrapper {
  margin: 0 auto;
  max-width: 800px;
}
