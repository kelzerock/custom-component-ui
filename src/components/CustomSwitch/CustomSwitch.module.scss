:root {
  --animation: .1s
}

.wrapper {
  display: inline-flex;
  gap: 5px;
  align-items: center;
  padding: 2px 4px;

  &:hover,
  &:hover .checkbox {
    cursor: pointer;
  }

  &.disabled:hover,
  &.disabled:hover .checkbox {
    cursor: default;
  }
}


.switchBox {
  width: 40px;
  height: 20px;
  position: relative;
  display: flex;
  justify-content: center;
  align-items: center;
}

.switch {
  display: block;
  width: 50%;
  height: 0px;
  overflow: visible;
  background-color: var(--color-main-blue);
  position: relative;
  z-index: 3;
}

.switchDash {
  height: 14px;
  border-radius: 10px;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  transform: translateY(-50%);
  background-color: var(--color-dark-gray);
  opacity: .5;
}

.checked .switchDash {
  background-color: inherit;
}

.switchPoint {
  height: 100%;
  aspect-ratio: 1;
  left: 0;
  transform: translateX(-50%);
  right: auto;
  position: relative;
  display: block;
  transition: left var(--animation) ease-in, transform var(--animation) ease-in;
  background-color: inherit;

  &::before {
    content: '';
    position: absolute;
    width: 20px;
    aspect-ratio: 1;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    background-color: inherit;
    border-radius: 50%;
    z-index: 3;
  }

  &::after {
    content: '';
    position: absolute;
    width: 30px;
    aspect-ratio: 1;
    left: 50%;
    top: 50%;
    transform: translate(-50%, -50%);
    background-color: inherit;
    border-radius: 50%;
    opacity: 0;
  }

  &:hover::after {
    opacity: .2;
  }
}

.checked .switchPoint {
  left: 100%;
  transform: translateX(-100%);
}

.checkbox {
  opacity: 0;
  width: 1px;
  height: 1px;
}

.title {
  padding: 2px;
}

.disabled .switchPoint {
  &:hover::after {
    opacity: 0;
  }
}

.disabled .title {
  color: var(--color-dark-gray-400);
}