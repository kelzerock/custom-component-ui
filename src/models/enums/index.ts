export { Checkbox } from "./enums-for-test/Checkbox";
