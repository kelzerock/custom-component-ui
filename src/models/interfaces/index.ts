export { CustomSelectProps } from "./custom-select-props";
export { CustomButtonProps } from "./custom-button-props";
export { CustomTextFieldProps } from "./custom-text-field-props";
export { OptionsData } from "./option-data";
