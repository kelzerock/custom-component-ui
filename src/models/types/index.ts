export { ButtonSize } from "./button-size";
export { ButtonVariant } from "./button-variant";
