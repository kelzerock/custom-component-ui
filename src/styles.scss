:root {
  --color-main-blue: #1976d2;
  --color-main-blue-100: #2070c0;
  --color-white: #fff;
  --color-white-100: #f6fafd;
  --color-red: #d32f2f;
  --color-gray: #f0f0f0;
  --color-dark-gray: #00000099;
}