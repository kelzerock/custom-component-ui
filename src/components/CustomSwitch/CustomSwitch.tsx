import { ChangeEvent, FC, useState } from "react";
import { Switch } from "../../models/enums";
import { CustomCheckboxProps } from "../../models/interfaces";
import styles from "./CustomSwitch.module.scss";
import clsx from "clsx";

export const CustomSwitch: FC<CustomCheckboxProps> = ({
  title,
  color,
  ...props
}: CustomCheckboxProps) => {
  const { required } = props;
  const requiredBlock = required ? " *" : "";
  const [checkedState, setCheckedState] = useState<boolean>(
    props.checked ?? false,
  );

  const handleClick = (event: ChangeEvent<HTMLInputElement>) => {
    if (props.disabled) return;

    if (props.onChange !== undefined) {
      props.onChange(event);
    } else {
      setCheckedState((prev) => !prev);
    }
  };

  const isControlled = props.checked !== undefined;
  const checked = isControlled ? props.checked : checkedState;

  return (
    <label
      className={clsx(
        styles.wrapper,
        props.disabled && styles.disabled,
        checked && styles.checked,
      )}
      data-testid={Switch.wrapper}
    >
      <span className={styles.switchBox} data-testid={Switch.switchBox}>
        <span
          className={styles.switch}
          style={{ backgroundColor: color }}
          data-testid={Switch.switch}
        >
          <span
            className={styles.switchDash}
            data-testid={Switch.switchDash}
          ></span>
          <span
            className={styles.switchPoint}
            data-testid={Switch.switchPoint}
          ></span>
        </span>
        <input
          onChange={handleClick}
          {...props}
          type="checkbox"
          checked={checked}
          className={styles.checkbox}
          data-testid={Switch.input}
        />
      </span>
      {title && (
        <span
          className={styles.title}
          data-testid={Switch.title}
        >{`${title}${requiredBlock}`}</span>
      )}
    </label>
  );
};
