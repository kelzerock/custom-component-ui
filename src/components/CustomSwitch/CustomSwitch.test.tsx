import { render, screen } from "@testing-library/react";
import { Switch } from "../../models/enums";
import styles from "./CustomSwitch.module.scss";
import userEvent from "@testing-library/user-event";
import { CustomSwitch } from "./CustomSwitch";

describe("Check CustomSwitch component", () => {
  it("All elements CustomSwitch rendered", () => {
    const titleTest = "text";
    render(<CustomSwitch title={titleTest} />);

    const wrapper = screen.getByTestId(Switch.wrapper);
    expect(wrapper).toBeInTheDocument();
    expect(wrapper).toHaveClass(styles.wrapper);

    const switchBox = screen.getByTestId(Switch.switchBox);
    expect(switchBox).toBeInTheDocument();
    expect(switchBox).toHaveClass(styles.switchBox);

    const switchTest = screen.getByTestId(Switch.switch);
    expect(switchTest).toBeInTheDocument();
    expect(switchTest).toHaveClass(styles.switch);

    const switchDash = screen.getByTestId(Switch.switchDash);
    expect(switchDash).toBeInTheDocument();
    expect(switchDash).toHaveClass(styles.switchDash);

    const switchPoint = screen.getByTestId(Switch.switchPoint);
    expect(switchPoint).toBeInTheDocument();
    expect(switchPoint).toHaveClass(styles.switchPoint);

    const checkbox = screen.getByTestId(Switch.input);
    expect(checkbox).toBeInTheDocument();
    expect(checkbox).toHaveClass(styles.checkbox);

    const title = screen.getByTestId(Switch.title);
    expect(title).toBeInTheDocument();
    expect(title).toHaveClass(styles.title);
  });

  it("Title doesn't rendered without props", () => {
    render(<CustomSwitch />);
    const title = screen.queryByTestId(Switch.title);
    expect(title).toBeNull();
  });

  it("Disable state correctly rendered", () => {
    render(<CustomSwitch disabled />);
    const wrapper = screen.getByTestId(Switch.wrapper);
    expect(wrapper).toBeInTheDocument();
    expect(wrapper).toHaveClass(styles.disabled);
    const checkbox = screen.getByTestId(Switch.input);
    expect(checkbox).toHaveAttribute("disabled");
  });

  it("Checked state correctly rendered when checked equal true", () => {
    render(<CustomSwitch checked />);
    const checkbox = screen.getByTestId(Switch.input);
    expect(checkbox).toHaveAttribute("checked");
  });

  it("Checked status correctly rendered when checked equal false", () => {
    render(<CustomSwitch checked={false} />);
    const checkbox = screen.getByTestId(Switch.input);
    expect(checkbox).not.toHaveAttribute("checked");
  });

  it("Correctly changed status checkbox", async () => {
    render(<CustomSwitch />);
    const checkbox = screen.getByTestId(Switch.input);
    expect(checkbox).toBeInTheDocument();
    const wrapper = screen.getByTestId(Switch.wrapper);
    expect(wrapper).toBeInTheDocument();

    if (!(checkbox instanceof HTMLInputElement)) {
      throw new Error("Checkbox is not an instance of HTMLInputElement");
    }

    expect(checkbox.checked).toBe(false);
    expect(wrapper).not.toHaveClass(styles.checked);
    await userEvent.click(checkbox);
    expect(checkbox.checked).toBe(true);
    expect(wrapper).toHaveClass(styles.checked);
    await userEvent.click(checkbox);
    expect(wrapper).not.toHaveClass(styles.checked);
    expect(checkbox.checked).toBe(false);
  });

  it("Checkbox's state toggles when clicking on label", async () => {
    render(<CustomSwitch title="Click me" />);
    const wrapper = screen.getByTestId(Switch.wrapper);
    const checkbox = screen.getByTestId(Switch.input);

    if (!(checkbox instanceof HTMLInputElement)) {
      throw new Error("Checkbox is not an instance of HTMLInputElement");
    }

    expect(checkbox.checked).toBe(false);
    await userEvent.click(wrapper);
    expect(checkbox.checked).toBe(true);
  });

  test("Check UI snapshot", () => {
    const { container } = render(<CustomSwitch title="Test" color="green" />);
    expect(container).toMatchSnapshot();
  });
});
