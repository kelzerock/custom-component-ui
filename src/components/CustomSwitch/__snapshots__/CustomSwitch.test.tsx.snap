// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check CustomSwitch component Check UI snapshot 1`] = `
<div>
  <label
    class="wrapper"
    data-testid="wrapper"
  >
    <span
      class="switchBox"
      data-testid="switchBox"
    >
      <span
        class="switch"
        data-testid="switch"
        style="background-color: green;"
      >
        <span
          class="switchDash"
          data-testid="switchDash"
        />
        <span
          class="switchPoint"
          data-testid="switchPoint"
        />
      </span>
      <input
        class="checkbox"
        data-testid="input"
        type="checkbox"
      />
    </span>
    <span
      class="title"
      data-testid="title"
    >
      Test
    </span>
  </label>
</div>
`;
